pkg/controller/externalnode/controller.go

```diff
@@ -0,0 +1,384 @@
+	// resourceVersion must be specified for update operation,
+	// so it gets the existing ExternalEntity and modifies the changed fields.
+	existingEE, _ := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Get(context.TODO(), ee.Name, metav1.GetOptions{})
+	isChanged := false
+	if !reflect.DeepEqual(existingEE.Spec, ee.Spec) {
+		existingEE.Spec = ee.Spec
+		isChanged = true
+	}
+	if !reflect.DeepEqual(existingEE.Labels, ee.Labels) {
+		existingEE.Labels = ee.Labels
+		isChanged = true
+	}
+	if isChanged {
+		_, err := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Update(context.TODO(), existingEE, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalNode(namespace string, name string) error {
+	obj, exists, _ := c.syncedExternalNode.GetByKey(k8s.NamespacedName(namespace, name))
+	if !exists {
+		klog.Warningf("Cannot find the ExternalNode namespace %s, name %s in the cache", namespace, name)
+		return nil
+	}
+	en := obj.(*v1alpha1.ExternalNode)
+	eeName := genExternalEntityName(en)
+	if eeName == "" {
+		klog.Warningf("Cannot find a network interface in ExternalNode namespace %s, name %s", namespace, name)
+		return nil
+	}
+	err := c.deleteExternalEntity(namespace, eeName)
+	if err != nil {
+		return err
+	}
+	c.syncedExternalNode.Delete(en)
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalEntity(namespace string, name string) error {
+	err := c.crdClient.CrdV1alpha2().ExternalEntities(namespace).Delete(context.TODO(), name, metav1.DeleteOptions{})
+	if errors.IsNotFound(err) {
+		klog.Warningf("Skipping deleting ExternalEntity namespace %s, name %s as it does not exist", namespace, name)
+		return nil
+	}
+	return err
+}
+
+func genExternalEntityName(externalNode *v1alpha1.ExternalNode) string {
+	if len(externalNode.Spec.Interfaces) == 0 {
+		return ""
+	}
+	// Only one network interface is supported now.
+	// Other interfaces except interfaces[0] will be ignored if there are more than one interfaces.
+	ifName := externalNode.Spec.Interfaces[0].Name
+	if ifName == "" {
+		return externalNode.Name
+	} else {
+		return externalNode.Name + "-" + ifName
+	}
+}
+
+func genExternalEntity(eeName string, en *v1alpha1.ExternalNode) *v1alpha2.ExternalEntity {
+	ownerRef := &metav1.OwnerReference{
+		APIVersion: "v1alpha1",
+		Kind:       "ExternalNode",
+		Name:       en.GetName(),
+		UID:        en.GetUID(),
+	}
+	endpoints := make([]v1alpha2.Endpoint, 0)
+	// Generate one/multiple endpoint(s) if one/multiple IP(s) are specified for interface[0].
+	// Generate one endpoint with name only if IP is not specified for interface[0].
+	if len(en.Spec.Interfaces[0].IPs) > 0 {
+		for _, ip := range en.Spec.Interfaces[0].IPs {
+			endpoints = append(endpoints, v1alpha2.Endpoint{
+				IP:   ip,
+				Name: en.Spec.Interfaces[0].Name,
+			})
+		}
+	} else {
+		endpoints = append(endpoints, v1alpha2.Endpoint{
```

Comment:
Why do we allow an ExternalEntity with empty IPs?

Reply:
My original thought is to support interface name alone, interface IP alone and both name and IPs. Do we prefer to let user always specify IP?

Reply:
For ExternalNode case, yes, I think IP should be required.

Reply:
OK, got it. Will update this part.